delta-lake/delta-33x/src/main/scala/com/nvidia/spark/rapids/delta/delta33x/GpuDelta33xParquetFileFormat.scala

```diff
@@ -273,6 +299,133 @@ case class GpuDelta33xParquetFileFormat(
         None
     }
   }
+}
+
+class DeltaMultiFileReaderFactory(
+    @transient sqlConf: SQLConf,
+    broadcastedConf: Broadcast[SerializableConfiguration],
+    dataSchema: StructType,
+    readDataSchema: StructType,
+    partitionSchema: StructType,
+    filters: Array[Filter],
+    @transient rapidsConf: RapidsConf,
+    metrics: Map[String, GpuMetric],
+    useMetadataRowIndex: Boolean,
+    tablePath: Option[String]
+    ) extends GpuParquetMultiFilePartitionReaderFactory(sqlConf, broadcastedConf,
+      dataSchema, readDataSchema, partitionSchema,
+      filters, rapidsConf, metrics, queryUsesInputFile = true) {
+
+  private val schemaWithIndices = readDataSchema.fields.zipWithIndex
+  def findColumn(name: String): Option[ColumnMetadata] = {
+    val results = schemaWithIndices.filter(_._1.name == name)
+    if (results.length > 1) {
+      throw new IllegalArgumentException(
+        s"There are more than one column with name=`$name` requested in the reader output")
+    }
+    results.headOption.map(e => ColumnMetadata(e._2, e._1))
+  }
+
+  private val isRowDeletedColumn = findColumn(IS_ROW_DELETED_COLUMN_NAME)
+  private val rowIndexColumnName = ROW_INDEX_COLUMN_NAME
+
+  private val rowIndexColumn = findColumn(rowIndexColumnName)
+
+  override def createColumnarReader(p: InputPartition): PartitionReader[ColumnarBatch] = {
+    val files = p.asInstanceOf[FilePartition].files
+    val reader = super.createColumnarReader(p)
+    new DeltaMultiFileParquetPartitionReader(files, reader,
+      isRowDeletedColumn, rowIndexColumn, broadcastedConf.value, tablePath, metrics)
+  }
+}
+
+class DeltaMultiFileParquetPartitionReader(
+    files: Array[PartitionedFile],
+    reader: PartitionReader[ColumnarBatch],
+    isRowDeletedColumnOpt: Option[ColumnMetadata],
+    rowIndexColumnOpt: Option[ColumnMetadata],
+    serializableConf: SerializableConfiguration,
+    tablePath: Option[String],
+    metrics: Map[String, GpuMetric]) extends PartitionReader[ColumnarBatch] {
+
+  private val filesMap = files.map(f => f.filePath.toString() -> f).toMap
+  private var file: PartitionedFile = null
+  private var rowIndex: Long = 0L
+  private var rowIndexFilterOpt: Option[RapidsRowIndexFilter] = None
+
+  override def next(): Boolean = {
+    reader.next()
+  }
+
+  override def close(): Unit = {
+    reader.close()
+  }
+
+  private def compareFile(file: PartitionedFile): Boolean = {
+    InputFileUtils.getCurInputFilePath() == file.urlEncodedPath &&
+      InputFileUtils.getCurInputFileStartOffset == file.start &&
+      InputFileUtils.getCurInputFileLength == file.length
+  }
+
+  override def get(): ColumnarBatch = {
+    val batch = reader.get()
+    if (isRowDeletedColumnOpt.isEmpty) {
+      // No columns to add, return the original batch
+      return batch
+    } else if (file == null || !compareFile(file)) {
```

Comment:
nit: could handle file == null in compareFile

Reply:
It's easier to read, though IMO.